main/src/ca/uwaterloo/flix/language/phase/Parser2.scala

```diff
@@ -529,6 +541,32 @@ object Parser2 {
   private val NAME_TAG = List(TokenKind.NameUpperCase, TokenKind.KeywordPure, TokenKind.KeywordImpure)
   private val NAME_PREDICATE = List(TokenKind.NameUpperCase)
 
+  private val MODIFIERS: List[TokenKind] = List(TokenKind.KeywordSealed, TokenKind.KeywordLawful, TokenKind.KeywordPub, TokenKind.KeywordInline, TokenKind.KeywordOverride)
+
+  /**
+    * Sets of the [[TokenKind]] that can appear as the first token of different concepts.
+    * For instance, FIRST_DECL are all the tokens that can be the first of a declaration ('enum', 'def', 'mod' and so on).
+    */
+  private val FIRST_DECL: List[TokenKind] = MODIFIERS ::: List(TokenKind.CommentDoc, TokenKind.Annotation, TokenKind.KeywordMod, TokenKind.KeywordDef, TokenKind.KeywordEnum, TokenKind.KeywordTrait, TokenKind.KeywordInstance, TokenKind.KeywordType, TokenKind.KeywordEff, TokenKind.KeywordRestrictable)
```

Comment:
I am debating if we should move these constants into the TokenKind class or their own class, instead of having them in the middle of the Parser.

Reply:
That's definitely another approach to encoding the information, just having a number of functions ala `isFirstType: TokenKind -> Boolean`. I think it makes sense to move to that actually...

Reply:
I forgot to answer your question
> Another question: Could e.g. FIRST_TYPE be a Boolean function on each token kind?

Yes. The FIRST set of a grammar rule is static data. The tokens that are possible as the first token in an expression for instance, does not change. I think this data belongs in the parser as `isFirst*` and `isRecover*` functions though. Let me try and argue that! 

To me the concerns we have are:
1. **Where is it practical to store FIRST and RECOVER sets?** I think in the parser, because if it ever needs to change that means the grammar is changing, and then you are already working in the parser. This is also the spot where I think you most often need to read FIRST and RECOVER sets.
2. **How can we have fast member checks on FIRST and RECOVER sets?** This is crucial for performance. By capturing the data as functions in the parser, we can change those functions in pursuit of performance, with no refactoring needed on the call-site.